python/ray/data/_internal/pipelined_dataset_iterator.py

```diff
@@ -0,0 +1,135 @@
+        batch_size: Optional[int] = 256,
+        batch_format: str = "default",
+        drop_last: bool = False,
+        local_shuffle_buffer_size: Optional[int] = None,
+        local_shuffle_seed: Optional[int] = None,
+    ) -> Iterator[DataBatch]:
+        ds = self._get_next_dataset()
+        return ds.iter_batches(
+            prefetch_blocks=prefetch_blocks,
+            batch_size=batch_size,
+            batch_format=batch_format,
+            drop_last=drop_last,
+            local_shuffle_buffer_size=local_shuffle_buffer_size,
+            local_shuffle_seed=local_shuffle_seed,
+        )
+
+    def iter_torch_batches(
+        self,
+        *,
+        prefetch_blocks: int = 0,
+        batch_size: Optional[int] = 256,
+        dtypes: Optional[Union["torch.dtype", Dict[str, "torch.dtype"]]] = None,
+        device: Optional[str] = None,
+        drop_last: bool = False,
+        local_shuffle_buffer_size: Optional[int] = None,
+        local_shuffle_seed: Optional[int] = None,
+    ) -> Iterator["TorchTensorBatchType"]:
+        from ray.air._internal.torch_utils import (
+            convert_ndarray_batch_to_torch_tensor_batch,
+        )
+
+        ds = self._get_next_dataset()
+        for batch in ds.iter_batches(
+            prefetch_blocks=prefetch_blocks,
+            batch_size=batch_size,
+            batch_format="numpy",
+            drop_last=drop_last,
+            local_shuffle_buffer_size=local_shuffle_buffer_size,
+            local_shuffle_seed=local_shuffle_seed,
+        ):
+            yield convert_ndarray_batch_to_torch_tensor_batch(
+                batch,
+                dtypes=dtypes,
+                device=device,
+            )
+
+    def to_tf(
+        self,
+        feature_columns: Union[str, List[str]],
+        label_columns: Union[str, List[str]],
+        *,
+        prefetch_blocks: int = 0,
+        batch_size: int = 1,
+        drop_last: bool = False,
+        local_shuffle_buffer_size: Optional[int] = None,
+        local_shuffle_seed: Optional[int] = None,
+    ) -> "tf.data.Dataset":
+        ds = self._get_next_dataset()
+        return Dataset.to_tf(
+            ds,
+            feature_columns,
+            label_columns,
+            prefetch_blocks=prefetch_blocks,
+            batch_size=batch_size,
+            drop_last=drop_last,
+            local_shuffle_buffer_size=local_shuffle_buffer_size,
+            local_shuffle_seed=local_shuffle_seed,
+        )
+
+    def stats(self) -> str:
+        return self._base_dataset_pipeline.stats()
+
+    def _with_backward_compat(self) -> DatasetIterator:
+        return PipelinedDatasetIteratorWithBackwardCompat(self)
+
+
+class PipelinedDatasetIteratorWithBackwardCompat(PipelinedDatasetIterator):
+    def __init__(
+        self,
+        dataset_iterator: PipelinedDatasetIterator,
+    ):
+        self._dataset_iterator = dataset_iterator
+
+    def __getattr__(self, name):
+        if name in "_dataset_iterator":
```

Comment:
Hmm this is a contains/substring check on `"_dataset_iterator"`, is that intentional or should this be an equality check?

Reply:
Ah should be equality check, nice catch.